src/glib/concurrent/thread.h

```diff
@@ -32,86 +32,92 @@ enum TMutexType {
+///   for a signal from the pool.
+///  Execute function represents the main functionality.
+///  IMPORTANT: thread pool takes ownership of runnable objects and will free them after a task
+///  executes them.
+class TThreadPool {
 public:
-	class TRunnable;
-		typedef TPt<TRunnable> PRunnable;
-	class TRunnable {
-	private:
-		TCRef CRef;
-	public:
-		friend class TPt<TRunnable>;
-	public:
-		TRunnable() {}
-		virtual ~TRunnable() {}
-
-		virtual void Run() = 0;
-		bool operator ==(const TRunnable& Other) const { return this == &Other; }
-	};
+    /// Has to implement Run()
+    class TRunnable {
+    public:
+        /// Emtpty default constructor
+        TRunnable() {}
+        /// Destructor
+        virtual ~TRunnable() {}
+        /// Main logic to be implemented
+        virtual void Run() = 0;
+    };
 private:
-	class TExecutorThread: public TThread {
-	private:
-		TThreadExecutor* Executor;
-		PRunnable Runnable;
-		PNotify Notify;
-	public:
-		TExecutorThread();
-		TExecutorThread(TThreadExecutor* Executor, const PNotify& Notify);
-
-		void Run();
-		void SetRunnable(const PRunnable& _Runnable) { Runnable = _Runnable; };
-	};
+    class TWorkerThread: public TThread {
+    private:
+        /// Pool is queried for new tasks that are to be run on this thread
+        TWPt<TThreadPool> Pool;
+        /// Notify
+        TWPt<TNotify> Notify;
+    public:
+        /// Only needed for Vector constructor
+        TWorkerThread();
+        /// Main constructor
+        TWorkerThread(TWPt<TThreadPool> Pool, const TWPt<TNotify>& Notify);
+        /// Waits for a task from the pool and executes it.
+        void Run();
+    };
 
 private:
-	typedef TThreadV<TExecutorThread> TExecThreadV;
-	typedef TLinkedQueue<PRunnable> TTaskQueue;
-
-	TExecThreadV ThreadV;
-	TTaskQueue TaskQ;
-
-	TCondVarLock Lock;
-
-	PNotify Notify;
-
-	volatile bool IsFinished;
-
+    /// Vector of threads
+    TVec<TWorkerThread> ThreadV;
+    /// Queue of tasks (runnable objects)
+    TLinkedQueue<TWPt<TRunnable>> TaskQ;
+    /// Used for signaling to worker threads when the queue is non-empty (work is to be done)
+    TCondVarLock Lock;
+    /// Notifiy
+    PNotify Notify;
 public:
-	TThreadExecutor(const TInt& PoolSize=1, const PNotify& Notify=TNullNotify::New());
-
-	~TThreadExecutor();
-
-	void Execute(const PRunnable& Runnable);
+    /// Creates and starts threads. When the pool is generated the threads
+    /// wait for the task queue to become non-empty.
+    TThreadPool(const int& PoolSize=1, const PNotify& Notify=TNullNotify::New());
+    /// Destructor waits for all work to stop
+    ~TThreadPool();
+    /// Pushes a runnable task to the queue, which will be executed when a thread is free.
+    /// IMPORTANT: takes ownership of Runnable and frees memmory when the
+    /// task is finished.
+    /// Sending NULL will stop one thread, which is used internally in destructor.
+    void Execute(const TWPt<TRunnable>& Runnable);
```

Comment:
I would make the Execute that accepts NULL private and add special Execute to public that checks for Runnable to not be NULL and passes it on to internal Execute. Unless there is use-case for non-private calls with NULL as parameter?